kythe/typescript/indexer.ts

```diff
@@ -1117,7 +1086,76 @@ class Visitor {
     if (decl.body) {
       this.visit(decl.body);
     } else {
-      if (kFunc) this.emitFact(kFunc, 'complete', 'incomplete');
+      this.emitFact(kFunc, 'complete', 'incomplete');
+    }
+  }
+
+  /**
+   * Visits a function parameters, which can be recursive in the case of
+   * parameters created via bound elements:
+   *    function foo({a, b: {c, d}}, e, f) {}
+   * In this code, a, c, d, e, f are all parameters with increasing parameter
+   * numbers [0, 4]. Ideally the parameter number would be hierarchical, with
+   * `a` having a parameter number of 0.0, but Kythe doesn't support that.
+   */
+  visitParameters(
+      parameters: ReadonlyArray<ts.ParameterDeclaration>, kFunc: VName) {
+    let paramNum = 0;
+    const recurseVisit =
+        (param: ts.ParameterDeclaration|ts.BindingElement) => {
+          this.visitDecorators(param.decorators || []);
+
+          switch (param.name.kind) {
+            case ts.SyntaxKind.Identifier:
+              const sym = this.getSymbolAtLocation(param.name);
+              if (!sym) {
+                this.todo(
+                    param.name, `param ${param.name.getText()} has no symbol`);
+                return;
+              }
+              const kParam = this.getSymbolName(sym, TSNamespace.VALUE);
+              this.emitNode(kParam, 'variable');
+
+              this.emitEdge(kFunc, `param.${paramNum}`, kParam);
+              ++paramNum;
+
+              if (ts.isParameterPropertyDeclaration(param)) {
+                // Class members defined in the parameters of a constructor are
+                // children of the class type.
+                const parentName = param.parent.parent.name;
+                if (parentName !== undefined) {
+                  const parentSym = this.getSymbolAtLocation(parentName);
+                  if (parentSym !== undefined) {
+                    const kClass =
+                        this.getSymbolName(parentSym, TSNamespace.TYPE);
+                    this.emitEdge(kParam, 'childof', kClass);
+                  }
+                }
+              } else {
+                this.emitEdge(kParam, 'childof', kFunc);
+              }
+
+              this.emitEdge(
+                  this.newAnchor(param.name), 'defines/binding', kParam);
+              break;
+            case ts.SyntaxKind.ObjectBindingPattern:
+            case ts.SyntaxKind.ArrayBindingPattern:
+              const elements = toArray(param.name.elements.entries());
+              for (const [index, element] of elements) {
+                if (ts.isBindingElement(element)) {
+                  recurseVisit(element);
+                }
+              }
+            default:
+              break;
+          }
+
+          if (ts.isParameter(param) && param.type) this.visitType(param.type);
+          if (param.initializer) this.visit(param.initializer);
+        }
+
+    for (const [, element] of toArray(parameters.entries())) {
```

Comment:
Isn't parameters already an array?  Why not
for (const element of parameters) {
?

Reply:
yeah, i dont know why i did this. thanks!